ostree/pull: Add network-retries command line option

rfairley opened the pull request:
This exposes a way to specify from the command line the number
of times to retry each download after a network error. If a negative
value is given, then the default number of retries (5) is used. If 0
is given, then errors are returned without retrying.

closes #1659

**Notes**
To-do:
- need to update bash-completion script and man page (will remove the "ME TODO" in the file once done)
- working on adding a test for this

Alternatively instead of initializing to -1, a `G_OPTION_ARG_CALLBACK` function could be used to 1) set a boolean variable to indicate that the `network-retries` argument was given and 2) set the `opt_network_retries` variable, and use the boolean variable to decide whether the option is added to the Variant. However, this may be complicating things.

jlebon commented:
Looks good so far! :+1: 

For tests, check out `test-pull-repeated.sh`. You can look at the original PR #1594 for the test case it added there to get familiar with it.

rfairley commented:
Thanks, taking a look at this now!

rfairley commented:
Just added the following:
- tests for `--network-retries` in `test-pull-repeated.sh`
- bash completion and `man` page updates
- moved `assert_fail` bash function from `test_remote_headers.sh` to `libtest.sh` (thinking that function could be used generally, please let me know if this should be kept in the original file)

jlebon commented on tests/test-pull-repeated.sh:
```diff
@@ -23,9 +23,10 @@ set -euo pipefail
+echo "1..4"
 
 COMMIT_SIGN="--gpg-homedir=${TEST_GPG_KEYHOME} --gpg-sign=${TEST_GPG_KEYID_1}"
+BIG_NUMBER_RETRIES=8 # Gives error rate <0.5%, with --random-408s=50
```
Minor: we don't have to make it a constant upfront. Let's just inline it where it's used? (Though the comment is definitely useful! Let's move that down there too!)

rfairley replied:
Makes sense, making this edit now. Thanks!
One commit plus a prep commit sounds good! Thank you for the link on rebasing, looking over this.

jlebon commented:
BTW, if you're not familiar with `git rebase -i`, it's definitely worth learning! Check out https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History.

rfairley commented:
Edits completed! Please let me know if there is anything else.

jlebon commented on tests/test-pull-repeated.sh:
```diff
@@ -67,3 +80,22 @@ ${CMD_PREFIX} ostree --repo=repo rev-parse main
+# Using 8 network retries gives error rate of <0.5%, when --random-408s=50
+if ${CMD_PREFIX} ostree --repo=repo pull --mirror origin --network-retries=8 main 2>err.txt; then
+   echo "Success with big number of network retries"
+fi
```
Minor: usually for tests we don't expect to fail, we just run it directly instead of wrapping it around an if statement. Because we're running with `set -e`, any command that fails will abort the test. See: http://redsymbol.net/articles/unofficial-bash-strict-mode/.

Not worth a respin, just FYI :)

jlebon replied:
Actually, thinking on this again, wrapping it around an if statement means that we'll silently continue even if it fails. So we do need to change this.

Want to post a follow-up PR once this one is merged? :)

rfairley replied:
Ah I see - previously I tried giving `--network-retries=0` to make it fail, and it did stop and fail, but for the wrong reason: it stopped at `${CMD_PREFIX} ostree --repo=repo rev-parse main` because nothing had been pulled in. Just tried running it directly now, it stopped at the pull command as it should. Will post a follow-up PR.

Thanks for pointing that out! Will read over the link.

jlebon commented:
Thanks for working on this!
@rh-atomic-bot r+ d01badc
